src/librustc/ty/mod.rs

```diff
@@ -1887,6 +1899,74 @@ impl<'a, 'gcx, 'tcx> AdtDef {
         debug!("sized_constraint_for_ty({:?}) = {:?}", ty, result);
         result
     }
+
+    fn dynsized_constraint_for_ty(&self,
+                                   tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                   ty: Ty<'tcx>)
+                                   -> Vec<Ty<'tcx>> {
+        let result = match ty.sty {
+            TyBool | TyChar | TyInt(..) | TyUint(..) | TyFloat(..) |
+            TyRawPtr(..) | TyRef(..) | TyFnDef(..) | TyFnPtr(_) |
+            TyStr | TyDynamic(..) | TySlice(_) | TyArray(..) |
+            TyTuple(..) | TyAdt(..) | TyClosure(..) | TyGenerator(..) |
```

Comment:
I don't understand the treatment of `TyAdt` here. Are we saying that e.g. a `struct` declaration is always `DynSized`?

Reply:
(Similarly `TyTuple`, `TyClosure` and `TyGenerator`...)

Reply:
@nikomatsakis I think that @plietar decided that `!DynSized` types should not be allowed in struct tails (in the initial implementation), so as a result of that, yes, all structs and tuples are `DynSized`. I can add a FIXME about that though, since it will be changed if `!DynSized` types are allowed in struct tails.

`TyClosure` and `TyGenerator` are currently always `Sized` (see `sized_constraints` above), so they are trivially also `DynSized`.

Reply:
OK. A FIXME would be nice, yes.